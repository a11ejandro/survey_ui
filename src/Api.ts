import axios, { AxiosResponse } from 'axios'
import { Statistics, Task } from './module'

const socketAddress = process.env.API_URL || ''
const axiosInstance = axios.create({
  baseURL: socketAddress + '/api/v1',
  timeout: 10000,
  headers: {
    common: {
      'Content-Type': 'application/json'
    }
  }
})


export async function createTask (task:Task) : Promise<AxiosResponse> {
  try {
    return await axiosInstance.post('tasks', task)
  } catch (e) {
    console.log(e)
    return Promise.reject(e)
  }
}

export async function getStatistics () : Promise<Statistics[]> {
  try {
    const response = await axiosInstance.get('statistics')
    return response.data as Statistics[]
  } catch (e) {
    console.log(e)
    return Promise.reject<Statistics[]>(e)
  }
}

