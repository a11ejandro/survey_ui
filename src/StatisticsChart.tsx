import * as React from 'react'
import 'chartjs-plugin-datalabels'
import { ChartData } from 'chart.js'
import { HorizontalBar } from 'react-chartjs-2'

import { Statistics } from './module'

interface Props {
  data?: Statistics[]
}

interface State {
  chartData: ChartData
}

const chartOptions = {
  plugins: {
    datalabels: {
      align: 'end',
      anchor: 'end',
      color: 'black'
    }
  }
}


export default class StatisticsChart extends React.Component<Props, State> {
  public static getDerivedStateFromProps(newProps: Props, oldState: State) {
    if (!newProps.data || newProps.data.length === 0 ) {
      return { chartData: null }
    }

    const labels:string[] = newProps.data.map((sample: Statistics) => sample.handler_type)
    const data:number[] = newProps.data.map((sample:Statistics) => sample.duration)

    const dataset = {
      label: newProps.data[0].collection_size + ' records',
      data
    }

    return {
      chartData: {
        labels,
        datasets: [dataset]
      }
    }
    
  }

  constructor (props:Props) {
    super(props)

    this.state = {} as State
  }

  public render () {
    if (!this.state.chartData) {
      return null
    }

    return (
      <HorizontalBar data={this.state.chartData} height={80} options={chartOptions}/>
    )
  }
}
