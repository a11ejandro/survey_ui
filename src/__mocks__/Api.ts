import { Statistics, Task } from '../module'

const promisify = (resolver: any) => {
  return new Promise((resolve, reject) => {
    resolve(resolver)
  })
}

export const statistics = [
  { collection_size: 1000, duration: 0.02, handler_type: 'go' },
  { collection_size: 1000, duration: 0.05, handler_type: 'ruby no AR'},
  { collection_size: 1000, duration: 0.1, handler_type: 'ruby + AR'}
]

export function getStatistics () {
  return promisify(statistics as Statistics[])
}

export function createTask (task: Task) {
  return promisify(true)
}
