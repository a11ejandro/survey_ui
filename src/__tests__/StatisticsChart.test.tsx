import * as React from 'react'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'

import { statistics } from '../__mocks__/Api'
import StatisticsChart from '../StatisticsChart'

let mountedComponent: any
const component = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<StatisticsChart data={statistics}/>)
  }
  return mountedComponent
}

const validData = {
  labels: ['go', 'ruby no AR', 'ruby + AR'],
  datasets: [
    {
      label: '1000 records',
      data: [0.02, 0.05, 0.1]
    }
  ]
}

it ('renders correctly', () => {
  expect(toJson(component())).toMatchSnapshot()
})

describe ('transforming data to chart data', () => {
  it ('transforms correct chart data', () => {
    expect(component().state().chartData).toEqual(validData)
  })

  it ('resets state when no data was passed', () => {
    component().setProps({data: []})
    expect(component().state().chartData).toBeNull()
  })
})

