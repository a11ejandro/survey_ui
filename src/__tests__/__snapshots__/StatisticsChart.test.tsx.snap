// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<HorizontalBar
  data={
    Object {
      "datasets": Array [
        Object {
          "data": Array [
            0.02,
            0.05,
            0.1,
          ],
          "label": "1000 records",
        },
      ],
      "labels": Array [
        "go",
        "ruby no AR",
        "ruby + AR",
      ],
    }
  }
  height={80}
  options={
    Object {
      "plugins": Object {
        "datalabels": Object {
          "align": "end",
          "anchor": "end",
          "color": "black",
        },
      },
    }
  }
/>
`;
